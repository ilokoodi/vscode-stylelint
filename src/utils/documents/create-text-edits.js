'use strict';

const diff = require('fast-diff');
const { Range, TextEdit } = require('vscode-languageserver-types');

/**
 * Creates text edits for a document given updated contents. Allows for
 * retaining the cursor position when the document is updated.
 * @param {lsp.TextDocument} document The document to create text edits for.
 * @param {string} newContents The new contents of the document.
 * @returns {lsp.TextEdit[]} The text edits to apply to the document.
 */
function createTextEdits(document, newContents) {
	const diffs = diff(document.getText(), newContents);

	const edits = [];
	let offset = 0;

	for (const [op, text] of diffs) {
		const start = offset;

		switch (op) {
			case diff.EQUAL:
				offset += text.length;
				break;
			case diff.DELETE:
				offset += text.length;
				edits.push(
					TextEdit.del(Range.create(document.positionAt(start), document.positionAt(offset))),
				);
				break;
			case diff.INSERT:
				edits.push(TextEdit.insert(document.positionAt(start), text));
				break;
		}
	}

	return edits;
}

module.exports = {
	createTextEdits,
};
