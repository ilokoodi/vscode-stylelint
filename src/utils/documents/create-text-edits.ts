import diff from 'fast-diff';
import { Range, TextEdit } from 'vscode-languageserver-types';
import { TextDocument } from 'vscode-languageserver-textdocument';

/**
 * Creates text edits for a document given updated contents. Allows for
 * retaining the cursor position when the document is updated.
 * @param document The document to create text edits for.
 * @param newContents The new contents of the document.
 * @returns The text edits to apply to the document.
 */
export function createTextEdits(document: TextDocument, newContents: string): TextEdit[] {
	const diffs = diff(document.getText(), newContents);

	const edits = [];
	let offset = 0;

	for (const [op, text] of diffs) {
		const start = offset;

		switch (op) {
			case diff.EQUAL:
				offset += text.length;
				break;
			case diff.DELETE:
				offset += text.length;
				edits.push(
					TextEdit.del(Range.create(document.positionAt(start), document.positionAt(offset))),
				);
				break;
			case diff.INSERT:
				edits.push(TextEdit.insert(document.positionAt(start), text));
				break;
		}
	}

	return edits;
}
