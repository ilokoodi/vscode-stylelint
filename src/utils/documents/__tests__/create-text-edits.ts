import { TextDocument } from 'vscode-languageserver-textdocument';

import { createTextEdits } from '../create-text-edits';

const createTextDocument = (text: string) =>
	TextDocument.create('file:///test.css', 'css', 1, text);

describe('createTextEdits', () => {
	test('should create text edits for insertions', () => {
		const textDocument = createTextDocument(`
			.foo {
				color: red;
			}
		`);
		const edits = createTextEdits(
			textDocument,
			`
			.foo {
				color: red;
				font-size: 12px;
			}
		`,
		);

		expect(edits).toMatchSnapshot();
	});

	test('should create text edits for deletions', () => {
		const textDocument = createTextDocument(`
			table.foo {
				width: 100%;
				height: 100%;
			}
		`);
		const edits = createTextEdits(
			textDocument,
			`
			table.foo {
				height: 100%;
			}
		`,
		);

		expect(edits).toMatchSnapshot();
	});

	test('should not create text edits without changes', () => {
		const textDocument = createTextDocument(`
			@media (min-width: 768px) {
				.foo {
					color: red;
				}
			}
		`);
		const edits = createTextEdits(
			textDocument,
			`
			@media (min-width: 768px) {
				.foo {
					color: red;
				}
			}
		`,
		);

		expect(edits).toEqual([]);
	});

	test('should create text edits for insertions and deletions', () => {
		const textDocument = createTextDocument(`
			a {
				font-size: 12px;
				font-weight: bold;
			}

			.foo {
				overflow: hidden;
				background: #fff url(/foo.png) no-repeat;
			}
		`);
		const edits = createTextEdits(
			textDocument,
			`
			a {
				font-size: 14px;
			}

			.foo {
				overflow: hidden;
				background: #e2e2e2 url('/bar.png') no-repeat;
			}
		`,
		);

		expect(edits).toMatchSnapshot();
	});
});
