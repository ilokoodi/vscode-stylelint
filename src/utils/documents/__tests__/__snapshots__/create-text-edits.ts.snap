// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTextEdits should create text edits for deletions 1`] = `
Array [
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 4,
        "line": 3,
      },
      "start": Object {
        "character": 4,
        "line": 2,
      },
    },
  },
]
`;

exports[`createTextEdits should create text edits for insertions 1`] = `
Array [
  Object {
    "newText": "	font-size: 12px;
			",
    "range": Object {
      "end": Object {
        "character": 3,
        "line": 3,
      },
      "start": Object {
        "character": 3,
        "line": 3,
      },
    },
  },
]
`;

exports[`createTextEdits should create text edits for insertions and deletions 1`] = `
Array [
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 17,
        "line": 2,
      },
      "start": Object {
        "character": 16,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "4",
    "range": Object {
      "end": Object {
        "character": 17,
        "line": 2,
      },
      "start": Object {
        "character": 17,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 21,
        "line": 3,
      },
      "start": Object {
        "character": 19,
        "line": 2,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 20,
        "line": 8,
      },
      "start": Object {
        "character": 17,
        "line": 8,
      },
    },
  },
  Object {
    "newText": "e2e2e2",
    "range": Object {
      "end": Object {
        "character": 20,
        "line": 8,
      },
      "start": Object {
        "character": 20,
        "line": 8,
      },
    },
  },
  Object {
    "newText": "'",
    "range": Object {
      "end": Object {
        "character": 25,
        "line": 8,
      },
      "start": Object {
        "character": 25,
        "line": 8,
      },
    },
  },
  Object {
    "newText": "",
    "range": Object {
      "end": Object {
        "character": 29,
        "line": 8,
      },
      "start": Object {
        "character": 26,
        "line": 8,
      },
    },
  },
  Object {
    "newText": "bar",
    "range": Object {
      "end": Object {
        "character": 29,
        "line": 8,
      },
      "start": Object {
        "character": 29,
        "line": 8,
      },
    },
  },
  Object {
    "newText": "'",
    "range": Object {
      "end": Object {
        "character": 33,
        "line": 8,
      },
      "start": Object {
        "character": 33,
        "line": 8,
      },
    },
  },
]
`;
